tidb-binlog/pump_client/pump.go

```diff
@@ -115,45 +129,35 @@ func (p *PumpStatus) createGrpcClient() error {
 	return nil
 }
 
-// ResetGrpcClient closes the pump's grpc connection.
-func (p *PumpStatus) ResetGrpcClient() {
-	p.Lock()
-	defer p.Unlock()
-
-	if p.grpcConn != nil {
-		p.grpcConn.Close()
-		p.Client = nil
-	}
-}
-
-// Reset resets the pump's grpc conn and err num.
+// Reset resets the pump's err num.
 func (p *PumpStatus) Reset() {
-	p.ResetGrpcClient()
 	atomic.StoreInt64(&p.ErrNum, 0)
 }
 
 // WriteBinlog write binlog by grpc client.
 func (p *PumpStatus) WriteBinlog(req *pb.WriteBinlogReq, timeout time.Duration) (*pb.WriteBinlogResp, error) {
 	p.RLock()
-	client := p.Client
-	p.RUnlock()
-
-	if client == nil {
+	if p.reCreateClient || p.grpcConn == nil {
+		p.RUnlock()
 		p.Lock()
-
-		if p.Client == nil {
-			if err := p.createGrpcClient(); err != nil {
+		if p.reCreateClient || p.grpcConn == nil {
+			p.reCreateClient = true
+			err := p.createGrpcClient()
+			if err != nil {
 				p.Unlock()
-				return nil, errors.Errorf("create grpc connection for pump %s failed, error %v", p.NodeID, err)
+				log.Info("[pumps client] write binlog to unavailable pump success, set this pump to avaliable", zap.String("NodeID", p.NodeID))
+				return nil, errors.Trace(err)
 			}
-		}
-		client = p.Client
 
-		p.Unlock()
+			p.Unlock()
```

Comment:
wrong `unlock`

Reply:
release the write lock and acquire the read lock again, when we create the `grpcConn` we hold the write lock, otherwise read lock to just call `WriteBinlog`